### PR TITLE
Replaces Java POJOs with Kotlin data classes

### LINKED ISSUE #4: Replace Java POJOs with Kotlin data classes
It will be nice to replace all those Java POJOs [in this package](https://github.com/YassinAJDI/PopularMovies/tree/master/app/src/main/java/com/ajdi/yassin/popularmovies/data/local/model) with Kotlin data classes: https://kotlinlang.org/docs/reference/data-classes.html

### PR DESCRIPTION
**Pull request by AniVerma17**
Resolves #4 . 	

**Primary changes**

- Replaces the POJOs defined for data models with equivalent Kotlin data classes

**Additional changes**

* Updates the app to target Android SDK version 29 (Android 10).
* Bug fixes for image loading with Glide.

### CODE REVIEW

**Comment by YassinAJDI** [comment]
Thanks for your contribution @AniVerma17 
I'm going to review this PR then merge it. Make sure you sign up for **hacktoberfest** [here](https://hacktoberfest.digitalocean.com/)

**Comment by AniVerma17** [comment]
> Thanks for your contribution @AniVerma17
> I'm going to review this PR then merge it. Make sure you sign up for **hacktoberfest** [here](https://hacktoberfest.digitalocean.com/)

@YassinAJDI I've already signed up. You can proceed with the PR.

File: app/src/main/java/com/ajdi/yassin/popularmovies/ui/moviedetails/trailers/TrailerViewHolder.java [hunk 1]
```diff
@@ -35,7 +35,7 @@ public TrailerViewHolder(@NonNull ItemTrailerBinding binding, Context context) {
     public void bindTo(final Trailer trailer) {
         String thumbnail =
                 "https://img.youtube.com/vi/" + trailer.getKey() + "/hqdefault.jpg";
-        GlideApp.with(context)
+        Glide.with(context)
```
Comment (shbhmnk) [inline app/src/main/java/com/ajdi/yassin/popularmovies/ui/moviedetails/trailers/TrailerViewHolder.java]: Does this improve anything? Please revert this change.

File: app/build.gradle [hunk 2]
```diff
@@ -1,13 +1,16 @@
 apply plugin: 'com.android.application'
+apply plugin: 'kotlin-android-extensions'
+apply plugin: 'kotlin-android'
+apply plugin: 'kotlin-kapt'
 
 android {
-    compileSdkVersion 28
+    compileSdkVersion 29
```
Comment (shbhmnk) [inline app/build.gradle]: This is not related to your PR. Could you please avoid such big changes in this PR.
This comment also applies to all the library changes that you made. Unless you really need them for updating POJO classes to Kotlin, I suggest you revert it.

**Review by shbhmnk — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/ajdi/yassin/popularmovies/ui/moviedetails/trailers/TrailerViewHolder.java (hunk 1 above)
Comment (AniVerma17) [reply app/src/main/java/com/ajdi/yassin/popularmovies/ui/moviedetails/trailers/TrailerViewHolder.java]: There is no `GlideApp` class defined in the project, thus it gives compile-time and build errors. Replacing it with the default Glide implementation fixed the problem.

File: app/build.gradle (hunk 2 above)
Comment (AniVerma17) [reply app/build.gradle]: Alright, I'll revert the versions. Rest of the library changes and additions are necessary to configure and use Kotlin in the project, so they are to be kept.

File: app/src/main/java/com/ajdi/yassin/popularmovies/ui/moviedetails/trailers/TrailerViewHolder.java (hunk 1 above)
Comment (shbhmnk) [reply app/src/main/java/com/ajdi/yassin/popularmovies/ui/moviedetails/trailers/TrailerViewHolder.java]: Did you successfully build the project before making any changes?

Comment (AniVerma17) [reply app/src/main/java/com/ajdi/yassin/popularmovies/ui/moviedetails/trailers/TrailerViewHolder.java]: Yes I did that, that's how I encountered this error.

File: app/build.gradle [hunk 3]
```diff
@@ -1,13 +1,16 @@
...
     defaultConfig {
         applicationId "com.ajdi.yassin.popularmovies"
         minSdkVersion 19
-        targetSdkVersion 28
-        versionCode 1
-        versionName "1.0"
+        targetSdkVersion 29
+        versionCode 2
+        versionName "1.1"
```
Comment (YassinAJDI) [inline app/build.gradle]: I agree with what @shubh10 said. 
No need here to upgrade the app version, unless database schema changed or new features been added.

File: app/build.gradle [hunk 4]
```diff
@@ -75,4 +78,10 @@ dependencies {
 
     // TextDrawable
     implementation 'com.amulyakhare:com.amulyakhare.textdrawable:1.0.1'
+    implementation "androidx.core:core-ktx:1.1.0"
+    implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:2.1.0"
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+}
+repositories {
+    mavenCentral()
```
Comment (YassinAJDI) [inline app/build.gradle]: Why do we need another repository in `build.gradle (app module)`? You can remove it.

File: app/src/main/java/com/ajdi/yassin/popularmovies/ui/moviedetails/trailers/TrailerViewHolder.java (hunk 1 above)
Comment (shbhmnk) [reply app/src/main/java/com/ajdi/yassin/popularmovies/ui/moviedetails/trailers/TrailerViewHolder.java]: `GlideApp` is a part of Glide's [Generated API](https://bumptech.github.io/glide/doc/generatedapi.html), please read the doc for more understanding.

Comment (AniVerma17) [reply app/src/main/java/com/ajdi/yassin/popularmovies/ui/moviedetails/trailers/TrailerViewHolder.java]: Never mind, I fixed it back. Just needed to switch the annotation processor for Glide to `kapt` .

Comment (shbhmnk) [reply app/src/main/java/com/ajdi/yassin/popularmovies/ui/moviedetails/trailers/TrailerViewHolder.java]: Sweet. Can you make sure to revert this change in other files too e.g. `BindingAdapters.java`.

File: app/build.gradle (hunk 2 above)
Comment (AniVerma17) [reply app/build.gradle]: Reverted `compileSdkVersion` and `targetSdkVersion` to 28.

File: app/src/main/java/com/ajdi/yassin/popularmovies/ui/moviedetails/trailers/TrailerViewHolder.java (hunk 1 above)
Comment (AniVerma17) [reply app/src/main/java/com/ajdi/yassin/popularmovies/ui/moviedetails/trailers/TrailerViewHolder.java]: Done.

**Review by shbhmnk — DISMISSED** [review]
Thanks for the changes.

File: app/build.gradle [hunk 5]
```diff
@@ -1,13 +1,16 @@
 apply plugin: 'com.android.application'
+apply plugin: 'kotlin-android-extensions'
+apply plugin: 'kotlin-android'
+apply plugin: 'kotlin-kapt'
```
Comment (YassinAJDI) [inline app/build.gradle]: - `'kotlin-android'` should be below `'com.android.application'`
- we use default annotation processing so remove `apply plugin: 'kotlin-kapt'`
- We are not using any of Kotlin extensions, so remove `apply plugin: 'kotlin-android-extensions'`

File: app/build.gradle [hunk 6]
```diff
@@ -75,4 +78,12 @@ dependencies {
 
     // TextDrawable
     implementation 'com.amulyakhare:com.amulyakhare.textdrawable:1.0.1'
+
+    //Kotlin dependencies
+    implementation "androidx.core:core-ktx:1.1.0"
+    implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:2.1.0"
```
Comment (YassinAJDI) [inline app/build.gradle]: we are not using Kotlin extensions yet. so remove both lines

File: app/build.gradle [hunk 7]
```diff
@@ -33,24 +36,24 @@ dependencies {
...
     implementation "androidx.recyclerview:recyclerview:$support_lib_version"
     implementation "androidx.cardview:cardview:$support_lib_version"
     implementation 'androidx.constraintlayout:constraintlayout:1.1.3'
 
     // Room
-    def room_version = "2.0.0"
+    def room_version = "2.1.0"
     implementation "androidx.room:room-runtime:$room_version"
-    annotationProcessor "androidx.room:room-compiler:$room_version"
+    kapt "androidx.room:room-compiler:$room_version"
```
Comment (YassinAJDI) [inline app/build.gradle:47]: Change to: `annotationProcessor "androidx.room:room-compiler:$room_version"`

**Review by YassinAJDI — CHANGES_REQUESTED** [review]
@AniVerma17 Please take a look at the small nitpicks, looks great otherwise!

File: app/build.gradle (hunk 3 above)
Comment (AniVerma17) [reply app/build.gradle]: Alright, I'll revert the app version.

File: app/build.gradle (hunk 4 above)
Comment (AniVerma17) [reply app/build.gradle]: This repository was automatically added when I configured the project for Kotlin. Anyway, the build runs fine without it, so I'll remove it.

File: app/build.gradle (hunk 5 above)
Comment (AniVerma17) [reply app/build.gradle]: > * `'kotlin-android'` should be below `'com.android.application'`

Alright, I'll reorder the code for this.

> * we use default annotation processing so remove `apply plugin: 'kotlin-kapt'`

`kapt` is required by the project since the code converted to Kotlin requires this to process the annotations for libraries like Room, Glide etc. So this is to be kept.

> * We are not using any of Kotlin extensions, so remove `apply plugin: 'kotlin-android-extensions'`

Yeah, this one was automatically added with Kotlin configuration. Not required now, so will remove this.

File: app/build.gradle (hunk 6 above)
Comment (AniVerma17) [reply app/build.gradle]: Yeah, these were automatically added with Kotlin configuration. Will remove these.

File: app/build.gradle (hunk 7 above)
Comment (AniVerma17) [reply app/build.gradle:47]: No, Room requires `kapt` for annotation processing when using Kotlin. The default `annotationProcessor` doesn't work with Kotlin code. So this is to be kept.

Comment (shbhmnk) [reply app/build.gradle:47]: @AniVerma17 Room's models are written in Java. Are you sure that room is asking for kapt?

Comment (AniVerma17) [reply app/build.gradle:47]: > @AniVerma17 Room's models are written in Java. Are you sure that room is asking for kapt?

They have been converted to Kotlin, since they were also among the POJO classes. So Room requires `kapt` for processing its annotations on those classes, else results in build error.

Comment (shbhmnk) [reply app/build.gradle:47]: You are right. Almost missed those files.
 :shipit:

File: app/build.gradle [hunk 8]
```diff
@@ -33,24 +35,24 @@ dependencies {
...
 
     // LiveData & ViewModel
-    def lifecycle_version = "2.0.0"
+    def lifecycle_version = "2.1.0"
     implementation "androidx.lifecycle:lifecycle-extensions:$lifecycle_version"
     annotationProcessor "androidx.lifecycle:lifecycle-compiler:$lifecycle_version"
```
Comment (YassinAJDI) [inline app/build.gradle]: Can you use `kapt `for this too?

**Review by YassinAJDI — COMMENTED** [review]
Thanks, LGTM. Just one minor change and it's ready for merge

File: app/build.gradle (hunk 7 above)
Comment (YassinAJDI) [reply app/build.gradle:47]: I tested on my local machine and didn't get any errors. I think because I already got those files in my cache.

File: app/build.gradle (hunk 8 above)
Comment (AniVerma17) [reply app/build.gradle]: Done.

**Review by YassinAJDI — APPROVED** [review]
Looking✨ to me! Thanks for your contribution.

**Review by ghost — COMMENTED** [review]
im sorry i didnt intend on hijacking your app im new

**Comment by ghost** [comment]
i only intended to get a repository

**Comment by YassinAJDI** [comment]
It's alright friend :)
